cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -1520,27 +1530,262 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
+                                PersistentVolumeClaim deletion = pvcOperations.get(namespace, pvcName);
+                                log.trace("Checking if {} {} in namespace {} has been deleted or recreated", pvc.getKind(), pvcName, namespace);
+                                return deletion == null || (deletion.getMetadata() != null && !uid.equals(deletion.getMetadata().getUid()));
+                            });
+
+                            waitForDeletionResults.add(waitForDeletion);
+                        }
+
+                        return CompositeFuture.join(waitForDeletionResults);
+                    })
+                    .compose(ignored -> {
+                        List<Future> createResults = new ArrayList<>(createPvcs.size());
+
+                        for (PersistentVolumeClaim pvc : createPvcs)    {
+                            log.debug("{}: Reconciling PVC {} for Pod {} after it was deleted and maybe recreated by the pod", reconciliation, pvc.getMetadata().getName(), podName);
+                            createResults.add(pvcOperations.reconcile(namespace, pvc.getMetadata().getName(), pvc));
+                        }
+
+                        return CompositeFuture.join(createResults);
+                    })
+                    .compose(ignored -> podOperations.readiness(namespace, podName, pollingIntervalMs, timeoutMs));
+
+            return fut;
+        }
+
+        /**
+         * Deletion of PVCs after the cluster is deleted is handled by owner reference and garbage collection. However,
+         * this would not help after scale-downs. Therefore we check if there are any PVCs which should not be present
+         * and delete them when they are.
+         *
+         * This should be called only after the Statefulset reconciliation, rolling update and scale-down when the PVCs
+         * are not used anymore by the pods.
+         *
+         * @return
+         */
+        Future<ReconciliationState> zkPersistentClaimDeletion() {
+            Future<ReconciliationState> futureResult = Future.future();
+            Future<List<PersistentVolumeClaim>> futurePvcs = pvcOperations.listAsync(namespace, Labels.fromMap(zkCluster.getSelectorLabels()));
+
+            futurePvcs.setHandler(res -> {
+                if (res.succeeded() && res.result() != null)    {
+                    List<String> maybeDeletePvcs = res.result().stream().map(pvc -> pvc.getMetadata().getName()).collect(Collectors.toList());
+                    List<String> desiredPvcs = zkCluster.generatePersistentVolumeClaims().stream().map(pvc -> pvc.getMetadata().getName()).collect(Collectors.toList());
+
+                    persistentClaimDeletion(maybeDeletePvcs, desiredPvcs).setHandler(futureResult.completer());
+                } else {
+                    futurePvcs.fail(res.cause());
```

Comment:
I think you mean to fail `futureResult` here, not `futurePvcs` which must have already been completed.

If the `res` succeeded but had null `result` theb we'd fail `futurePvcs` here with a null cause. I suspect that's not really what you want. I don't think `res.result()` can actually be null, so probably all you need to do it remove the `&& res.result() != null` from the `if` clause.